bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/registry/AbstractRegistry.java

```diff
@@ -268,6 +279,15 @@ protected void addProvider(Provider<E> provider) {
         }
     }
 
+    /**
+     * Returns the key for a given element
+     *
+     * @param element
+     *            element
+     * @return key (must not be null)
+     */
+    protected abstract K getKey(E element);
```

Comment:
Hm, that's getting more complicated than I had hoped. Looping through all entries and figuring out the key for every element looks pretty inefficient to me.
The underlying problem is that our `Provider<E>` interface does not at all HAVE the notion of a key - this only exists for `ManagedProvider`.
So by definition, we do not have a key at hand for the provided entities (and we do not even know, if such a key exists). I think our best option is the equals method - if an entity has a key, this should be the one that is used in equals; for our `Rule` class, this is e.g. the case.
Wdyt?